iup/__init__.py

```diff
@@ -831,3 +846,446 @@ def predict(
+
+    mean_bias = sum of the bias across time/length of data
+    -------------------------
+    Input: data, pred in a matching and validated format
+    Return: pl.DataFrame with mean bias and the forecast start date and end date
+    """
+
+    # Check the conditions for date match:
+    # 1. Mutual dates must exist between data and prediction.
+    assert any(
+        data["date"].is_in(pred["date"])
+    ), "No matched dates between data and prediction."
+
+    # 2. There should not be any duplicated date in either data or prediction.
+    common_dates = data.filter(pl.col("date").is_in(pred["date"])).select("date")
+
+    assert (
+        len(common_dates) == common_dates.n_unique()
+    ), "Duplicated dates are found in data or prediction."
+
+    joined = (
+        data.join(pred, on="date", how="inner", validate="1:1")
+        .rename({"estimate": "data", "estimate_right": "pred"})
+        .with_columns(diff=(pl.col("data") - pl.col("pred")))
+    )
+
+    joined = joined.with_columns(bias=joined["diff"].sign())
+
+    m_bias = pl.DataFrame(
+        {
+            "forecast_start": joined["date"].min(),
+            "forecast_end": joined["date"].max(),
+            "mbias": joined["bias"].sum() / joined.shape[0],
+        }
+    )
+
+    return m_bias
+
+
+def get_eos_abe(data: IncidentUptakeData, pred: PointForecast) -> pl.DataFrame:
+    """
+    Calculate the absolute error of the total uptake at the end of season between data and prediction.
+    -------------------
+    Input: data, pred in a matching and validated format
+    Return: pl.DataFrame with absolute error in the total uptake between data and prediction
+            and the forecast end date.
+    """
+    joined = (
+        data.join(pred, on="date", how="inner", validate="1:1")
+        .rename({"estimate": "data", "estimate_right": "pred"})
+        .with_columns(
+            cumu_data=pl.col("data").cum_sum(),
+            cumu_pred=pl.col("pred").cum_sum(),
+        )
+        .filter(pl.col("date") == pl.col("date").max())
+        .rename({"date": "forecast_end"})
+    )
+
+    abe_perc = abs(joined["cumu_data"] - joined["cumu_pred"]) / joined["cumu_data"]
+
+    return pl.DataFrame([joined["forecast_end"], abe_perc])
+
+
+# #### evaluation metrics #####
+# class PointMetric(pl.DataFrame, metaclass=abc.ABCMeta):
+#     """
+#     Abstract class for evaluation metrics for point estimate forecast.
+#     """
+
+#     def __init__(self, data: PointForecast, pred: PointForecast):
+#         super().__init__(self.preprocess(data, pred))
+
+#     """
+#     Combine data and prediction together, varied by metric type (time-wise or not)
+#     """
+
+#     @abc.abstractmethod
+#     def preprocess(self, data, pred):
+#         pass
+
+#     """
+#     Evaluation metric, varied by metric type.
+#     """
+
+#     @abc.abstractmethod
+#     def get_metric(self, metric_type):
+#         pass
+
+
+# # Any metric that does not require time-wise matching, can go here.
```

Comment:
I haven't read through anything below this point thoroughly, since it is all commented out. Does that mean you intend to save it for future reference? If so, I recommend you delete it here and stash it elsewhere, to try to keep __init__.py as clean as possible.

Reply:
Sure! Sorry for the redundancy.

Reply:
test